automata/pda/pda.py

```diff
@@ -24,6 +47,313 @@ class PDA(Automaton, metaclass=abc.ABCMeta):
+            + ("ε" if stack_push == "" else str(stack_push))
+        )
+
+    @staticmethod
+    def _get_symbol_configuration(
+        from_state: PDAConfiguration, to_state: PDAConfiguration
+    ) -> Tuple[str, str, str]:
+        if (
+            from_state.remaining_input == to_state.remaining_input
+            or len(from_state.remaining_input) == 0
+        ):
+            input_symbol = ""
+        else:
+            input_symbol = from_state.remaining_input[0]
+
+        stack_top_symbol = from_state.stack.top()
+
+        if len(from_state.stack) == len(to_state.stack) + 1:
+            stack_push_symbols = ""
+        else:
+            stack_push_symbols = "".join(to_state.stack[len(from_state.stack) - 1 :])
+
+        return (input_symbol, stack_top_symbol, stack_push_symbols[::-1])
+
+    @abc.abstractmethod
+    def iter_transitions(
+        self,
+    ) -> Generator[Tuple[PDAStateT, PDAStateT, Tuple[str, str, str]], None, None]:
+        """
+        Iterate over all transitions in the automaton.
+        Each transition is a tuple of the form
+        (from_state, to_state, (input_symbol, stack_top_symbol, stack_push_symbols))
+        """
+
+    def show_diagram(
+        self,
+        input_str: Optional[str] = None,
+        with_machine: bool = True,
+        with_stack: bool = True,
+        path: Union[str, os.PathLike, None] = None,
+        *,
+        layout_method: LayoutMethod = "dot",
+        horizontal: bool = True,
+        reverse_orientation: bool = False,
+        fig_size: Union[Tuple[float, float], Tuple[float], None] = None,
+        font_size: float = 14.0,
+        arrow_size: float = 0.85,
+        state_separation: float = 0.5,
+    ) -> pgv.AGraph:
+        """
+        Generates the graph associated with the given PDA.
+        Args:
+            - input_str (str, optional): String list of input symbols. Defaults to None.
+            - with_machine (bool, optional): Constructs the diagram with states and
+              transitions. Ignored if `input_str` is None. Default to True.
+            - with_stack (bool, optional): Constructs the diagram with stack and its
+              operations. Ignored if `input_str` is None. Default to True.
+            - path (str or os.PathLike, optional): Path to output file. If
+              None, the output will not be saved.
+            - horizontal (bool, optional): Direction of node layout. Defaults
+              to True.
+            - reverse_orientation (bool, optional): Reverse direction of node
+              layout. Defaults to False.
+            - fig_size (tuple, optional): Figure size. Defaults to None.
+            - font_size (float, optional): Font size. Defaults to 14.0.
+            - arrow_size (float, optional): Arrow head size. Defaults to 0.85.
+            - state_separation (float, optional): Node distance. Defaults to 0.5.
+        Returns:
+            AGraph corresponding to the given automaton.
+        """
+
```

Comment:
See previous comment, should add an error message for missing imports.

Reply:
The first line in this function calls `create_graph` which takes care of raising the correct error message.

If you want me to add it to this function anyway, I can do it. Let me know.

Reply:
It looks like `create_graph` only raises an error message if `pygraphviz` is missing, but this function uses `coloraide` as well, so I think we need the additional check.